### DISCUSSION TITLE
`quantity[Int, A]` extension on Iterator[A], `quantity[Long, A]` on a F[A]: Foldable ?

### DISCUSSION THREAD

**Discussion by benhutchison** [erikerlandson/coulomb#436] [General]
If our [Data Types are units](http://erikerlandson.github.io/blog/2020/04/26/your-data-type-is-a-unit/), then it would seem that a collection's `size` ought to yield a typed `Quantity`. A `quantity` extension method could be offered over sized-collections eg
- `Iterator` and the Scala collections API
- Cats Foldable

These would presumably be in separate modules since they are distinct from other Coulomb features

**Comment by erikerlandson** [comment]
@benhutchison  yes, this has been on my to-do list for a long time.  In scala 3 they redesigned the "internals" of scala's collection library, and I need to get familiar with how these all work to best integrate with the collections lib.

xref: https://github.com/erikerlandson/coulomb/issues/50

But yes, there should be a `.qlength` or `.ulength` that returns `Quantity[Int, A]` for any `Seq[A]` - this bit would be easy, just an `extension` method.

**Comment by benhutchison** [comment]
Some sort of `qsum` operation for Collections of Quantitys would also be useful. The existing `sum` requires a `Numeric` and it isn't possible to implement  `Numeric[Quantity[V, U]]` (Yes I tried :P).

**Comment by benhutchison** [comment]
Actually @erikerlandson , when I tried defining a `qsum: Quantity[V, U]` over an `Iterable[Quantity[V, U]]`, I ran into the problem of needing a `zero: Quantity[V, U]` operation to begin the sum fold with. 

Have you thought about providing a `zero` (and possibly `one`) type-class for quantities?

**Reply by erikerlandson** [reply]
Yes, `Numeric` isn't possible, quantities aren't type-closed under multiplication.  I think you could get away with defining `AdditiveGroup`, which includes a zero.

Ironically, `RuntimeQuantity` would work more easily here since the unit type is `RuntimeUnit` and it has no type closure problems.
